tests/components/roku/test_config_flow.py

```diff
@@ -0,0 +1,247 @@
+from homeassistant.components.ssdp import (
+    ATTR_SSDP_LOCATION,
+    ATTR_UPNP_FRIENDLY_NAME,
+    ATTR_UPNP_SERIAL,
+)
+from homeassistant.config_entries import SOURCE_IMPORT, SOURCE_SSDP, SOURCE_USER
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_SOURCE
+from homeassistant.data_entry_flow import (
+    RESULT_TYPE_ABORT,
+    RESULT_TYPE_CREATE_ENTRY,
+    RESULT_TYPE_FORM,
+)
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.setup import async_setup_component
+
+from tests.components.roku import (
+    HOST,
+    SSDP_LOCATION,
+    UPNP_FRIENDLY_NAME,
+    UPNP_SERIAL,
+    MockDeviceInfo,
+    setup_integration,
+)
+
+
+async def async_configure_flow(
+    hass: HomeAssistantType, flow_id: str, user_input: Optional[Dict] = None
+) -> Any:
+    """Set up mock Roku integration flow."""
+    with patch(
+        "homeassistant.components.roku.config_flow.Roku.device_info",
+        new=MockDeviceInfo,
+    ):
+        return await hass.config_entries.flow.async_configure(
+            flow_id=flow_id, user_input=user_input
+        )
+
+
+async def async_init_flow(
+    hass: HomeAssistantType,
+    handler: str = DOMAIN,
+    context: Optional[Dict] = None,
+    data: Any = None,
+) -> Any:
+    """Set up mock Roku integration flow."""
+    with patch(
+        "homeassistant.components.roku.config_flow.Roku.device_info",
+        new=MockDeviceInfo,
+    ):
+        return await hass.config_entries.flow.async_init(
+            handler=handler, context=context, data=data
+        )
+
+
+async def test_duplicate_error(hass: HomeAssistantType) -> None:
+    """Test that errors are shown when duplicates are added."""
+    await setup_integration(hass, skip_entry_setup=True)
+
+    result = await async_init_flow(
+        hass, context={CONF_SOURCE: SOURCE_IMPORT}, data={CONF_HOST: HOST}
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+    result = await async_init_flow(
+        hass, context={CONF_SOURCE: SOURCE_USER}, data={CONF_HOST: HOST}
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+    result = await async_init_flow(
+        hass,
+        context={CONF_SOURCE: SOURCE_SSDP},
+        data={
+            ATTR_UPNP_FRIENDLY_NAME: UPNP_FRIENDLY_NAME,
+            ATTR_SSDP_LOCATION: SSDP_LOCATION,
+            ATTR_UPNP_SERIAL: UPNP_SERIAL,
+        },
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+
+async def test_form(hass: HomeAssistantType) -> None:
+    """Test the user step."""
+    await async_setup_component(hass, "persistent_notification", {})
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
+    )
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] == {}
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5184723)